File: src/API/AdventureWorks.Contracts/v1/Common/Responses/PagedResponse.cs
using System.Text.Json.Serialization;

namespace AdventureWorks.Contracts.v1.Common.Responses;

public class PagedResponse<TResponse>
{
    [JsonPropertyOrder(1)]
    public required int Page { get; init; }
    
    [JsonPropertyOrder(2)]
    public required int PageSize { get; init; }
    
    [JsonPropertyOrder(3)]
    public required int? TotalCount { get; init; }
    

    [JsonPropertyOrder(9)]
    public bool? HasNextPage => TotalCount is null ? null : TotalCount > (Page * PageSize);
    
    [JsonPropertyOrder(10)]

    public required IEnumerable<TResponse> Items { get; init; } = Enumerable.Empty<TResponse>();
}

File: src/UI/src/types/pagedResponse.ts
import { z } from 'zod';

export const pagedResponseSchema = <T extends z.ZodTypeAny>(itemSchema: T) =>
  z.object({
    page: z.number(),
    pageSize: z.number(),
    totalCount: z.number().nullable(),
    hasNextPage: z.boolean().nullable(),
    items: z.array(itemSchema)
  });

export type PagedResponse<TResponse> = z.infer<ReturnType<typeof pagedResponseSchema<z.ZodType<TResponse>>>>;
